File: Airship/AirshipCore/Source/common/UACustomEvent.m
/* Copyright Airship and Contributors */

#import "UACustomEvent+Internal.h"
#import "UAAnalytics.h"
#import "UAirship.h"
#import "NSJSONSerialization+UAAdditions.h"

@implementation UACustomEvent

NSString * const UAInteractionMCRAP = @"ua_mcrap";
const NSUInteger UACustomEventCharacterLimit = 255;
const NSUInteger UACustomEventMaxPropertiesCount = 100;
const NSUInteger UACustomEventMaxPropertyCollectionSize = 20;
const NSUInteger UACustomEventMaxPropertiesSize = 65536;

// Public data keys
NSString *const UACustomEventNameKey = @"event_name";
NSString *const UACustomEventValueKey = @"event_value";
NSString *const UACustomEventPropertiesKey = @"properties";
NSString *const UACustomEventTransactionIDKey = @"transaction_id";
NSString *const UACustomEventInteractionIDKey = @"interaction_id";
NSString *const UACustomEventInteractionTypeKey = @"interaction_type";

// Private data keys
NSString *const UACustomEventConversionMetadataKey = @"conversion_metadata";
NSString *const UACustomEventConversionSendIDKey = @"conversion_send_id";
NSString *const UACustomEventTemplateTypeKey = @"template_type";

- (NSString *)eventType {
    return @"enhanced_custom_event";
}

- (instancetype)initWithName:(NSString *)eventName withValue:(NSDecimalNumber *)eventValue {
    self = [super init];
    if (self) {
        self.eventName = eventName;
        self.eventValue = eventValue;
        self.properties = [NSDictionary dictionary];
    }

    return self;
}

+ (instancetype)eventWithName:(NSString *)eventName {
    return [self eventWithName:eventName value:nil];
}

+ (instancetype)eventWithName:(NSString *)eventName valueFromString:(NSString *)eventValue {
    NSDecimalNumber *decimalValue = eventValue ? [NSDecimalNumber decimalNumberWithString:eventValue] : nil;
    return [self eventWithName:eventName value:decimalValue];
}

+ (instancetype)eventWithName:(NSString *)eventName value:(NSDecimalNumber *)eventValue {
    return [[self alloc] initWithName:eventName withValue:eventValue];
}

- (void)setBoolProperty:(BOOL)value forKey:(NSString *)key {
    NSMutableDictionary *mutableProperties = self.properties.mutableCopy;
    [mutableProperties setValue:@(value) forKey:key];
    self.properties = mutableProperties.copy;
}

- (void)setStringProperty:(NSString *)value forKey:(NSString *)key {
    NSMutableDictionary *mutableProperties = self.properties.mutableCopy;
    [mutableProperties setValue:[value copy] forKey:key];
    self.properties = mutableProperties.copy;
}

- (void)setNumberProperty:(NSNumber *)value forKey:(NSString *)key {
    NSMutableDictionary *mutableProperties = self.properties.mutableCopy;
    [mutableProperties setValue:[value copy] forKey:key];
    self.properties = mutableProperties.copy;
}

- (void)setStringArrayProperty:(NSArray *)value forKey:(NSString *)key {
    NSMutableDictionary *mutableProperties = self.properties.mutableCopy;
    [mutableProperties setValue:[value copy] forKey:key];
    self.properties = mutableProperties.copy;
}

- (void)setEventValue:(NSDecimalNumber *)eventValue {
    if (!eventValue) {
        _eventValue = nil;
    } else {
        if ([eventValue isKindOfClass:[NSDecimalNumber class]]) {
            _eventValue = eventValue;
        } else {
            _eventValue = [NSDecimalNumber decimalNumberWithDecimal:[eventValue decimalValue]];
        }
    }
}


- (BOOL)isValid {
    BOOL isValid = YES;

    if (!self.eventName.length || self.eventName.length > UACustomEventCharacterLimit) {
        UA_LERR(@"Event name must be between 1 and %lu characters.", (unsigned long)UACustomEventCharacterLimit);
        isValid = NO;
    }

    if (self.interactionType.length > UACustomEventCharacterLimit) {
        UA_LERR(@"Event interaction type is larger than %lu characters.", (unsigned long)UACustomEventCharacterLimit);
        isValid = NO;
    }

    if (self.interactionID.length > UACustomEventCharacterLimit) {
        UA_LERR(@"Event interaction ID is larger than %lu characters.", (unsigned long)UACustomEventCharacterLimit);
        isValid = NO;
    }

    if (self.transactionID.length > UACustomEventCharacterLimit) {
        UA_LERR(@"Event transaction ID is larger than %lu characters.", (unsigned long)UACustomEventCharacterLimit);
        isValid = NO;
    }

    if (self.templateType.length > UACustomEventCharacterLimit) {
        UA_LERR(@"Event template type is larger than %lu characters.", (unsigned long)UACustomEventCharacterLimit);
        isValid = NO;
    }

    if (self.eventValue) {
        if ([self.eventValue isEqualToNumber:[NSDecimalNumber notANumber]]) {\
            UA_LERR(@"Event value is not a number.");
            isValid = NO;
        } else if ([self.eventValue compare:@(INT32_MAX)] > 0) {
            UA_LERR(@"Event value %@ is larger than 2^31-1.", self.eventValue);
            isValid = NO;
        } else if ([self.eventValue compare:@(INT32_MIN)] < 0) {
            UA_LERR(@"Event value %@ is smaller than -2^31.", self.eventValue);
            isValid = NO;
        }
    }
    
    NSError *error;
    NSData *propertyData = [NSJSONSerialization dataWithJSONObject:self.properties options:0 error:&error];
  
    if (error) {
        UA_LERR(@"Event properties serialization error %@", error);
        isValid = NO;
    } else if (propertyData.length > UACustomEventMaxPropertiesSize) {
        UA_LERR(@"Event properties (%lu bytes) are larger than the maximum size of %lu bytes.", (unsigned long) propertyData.length, (unsigned long)UACustomEventMaxPropertiesSize);
        isValid = NO;
    }

    return isValid;
}

- (void)setInteractionFromMessageCenterMessage:(NSString *)messageID {
    self.interactionID = messageID;
    self.interactionType = UAInteractionMCRAP;
}

- (NSDictionary *)data {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];

    // Event name
    [dictionary setValue:self.eventName forKey:UACustomEventNameKey];

    // Conversion Send ID
    NSString *sendID = self.conversionSendID ?:[UAirship analytics].conversionSendID;
    [dictionary setValue:sendID forKey:UACustomEventConversionSendIDKey];

    // Conversion Send Metadata
    NSString *sendMetadata = self.conversionPushMetadata ?:[UAirship analytics].conversionPushMetadata;
    [dictionary setValue:sendMetadata forKey:UACustomEventConversionMetadataKey];

    // Interaction
    [dictionary setValue:self.interactionID forKey:UACustomEventInteractionIDKey];
    [dictionary setValue:self.interactionType forKey:UACustomEventInteractionTypeKey];

    // Transaction ID
    [dictionary setValue:self.transactionID forKey:UACustomEventTransactionIDKey];

    // Template type
    [dictionary setValue:self.templateType forKey:UACustomEventTemplateTypeKey];

    // Event value
    if (self.eventValue) {

        // Move the decimal position over 6 positions
        NSDecimalNumber *number = [self.eventValue decimalNumberByMultiplyingByPowerOf10:6];

        /*
         We use long long value here because on a 32 bit machine int and long
         have the same range. We clamp eventValue to [-2^31, 2^31-1] so we know
         for sure that the value multiplied by 10^6 (~2^20) will have an approximate
         range of [-2^51, 2^51-1] so it will always fit into the range warp9
         is expecting [-2^63, 2^63-1].
         */
        [dictionary setValue:@([number longLongValue]) forKey:UACustomEventValueKey];
    }

    // Properties
    
    [dictionary setValue:self.properties forKey:UACustomEventPropertiesKey];

    return dictionary.copy;
}

- (NSDictionary *)payload {
    /*
     * We are unable to use the event.data for automation because we modify some
     * values to be stringified versions before we store the event to be sent to
     * warp9. Instead we are going to recreate the event data with the unmodified
     * values.
     */
    NSDictionary *eventData = [NSMutableDictionary dictionary];
    [eventData setValue:self.eventName forKey:UACustomEventNameKey];
    [eventData setValue:self.interactionID forKey:UACustomEventInteractionIDKey];
    [eventData setValue:self.interactionType forKey:UACustomEventInteractionTypeKey];
    [eventData setValue:self.transactionID forKey:UACustomEventTransactionIDKey];
    [eventData setValue:self.eventValue forKey:UACustomEventValueKey];
    [eventData setValue:self.properties forKey:UACustomEventPropertiesKey];

    return eventData;
}

- (void)track {
    [[UAirship analytics] addEvent:self];
}

@end

File: Airship/AirshipCore/Source/UACustomEvent.swift
/* Copyright Airship and Contributors */

/**
 * A UACustomEvent captures information regarding a custom event for
 * UAAnalytics.
 */
@objc
public class UACustomEvent : NSObject, UAEvent {

    private static let interactionMCRAP = "ua_mcrap"

    /**
     * The max number of properties.
     */
    @objc
    public static let maxPropertiesSize = 65536

    // Public data keys
    @objc
    public static let eventNameKey = "event_name"
    @objc
    public static let eventValueKey = "event_value"
    @objc
    public static let eventPropertiesKey = "properties"
    @objc
    public static let eventTransactionIDKey = "transaction_id"
    @objc
    public static let eventInteractionIDKey = "interaction_id"
    @objc
    public static let eventInteractionTypeKey = "interaction_type"

    // Private data keys
    static let eventConversionMetadataKey = "conversion_metadata"
    static let eventConversionSendIDKey = "conversion_send_id"
    static let eventTemplateTypeKey = "template_type"

    /**
     * The send ID that triggered the event.
     * @note For internal use only. :nodoc:
     */
    @objc
    public var conversionSendID: String?

    /**
     * The conversion push metadata.
     * @note For internal use only. :nodoc:
     */
    @objc
    public var conversionPushMetadata: String?

    /**
     * The event's template type. The template type's length must not exceed 255 characters or it will
     * invalidate the event.
     * @note For internal use only. :nodoc:
     */
    @objc
    public var templateType: String?

    private var _eventValue : NSDecimalNumber?

    /**
     * The event's value. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     */
    @objc
    public var eventValue : NSNumber? {
        get {
            return self._eventValue
        }
        set {
            if let newValue = newValue {
                if let decimal = newValue as? NSDecimalNumber  {
                    self._eventValue = decimal
                } else {
                    let converted = NSDecimalNumber.init(decimal: newValue.decimalValue)
                    self._eventValue = converted
                }
            } else {
                self._eventValue = nil
            }
        }
    }

    /**
     * The event's name. The name's length must not exceed 255 characters or it will
     * invalidate the event.
     */
    @objc
    public var eventName: String?

    /**
     * The event's transaction ID. The ID's length must not exceed 255 characters or it will
     * invalidate the event.
     */
    @objc
    public var transactionID: String?

    /**
     * The event's interaction type. The type's length must not exceed 255 characters or it will
     * invalidate the event.
     */
    @objc
    public var interactionType: String?

    /**
     * The event's interaction ID. The ID's length must not exceed 255 characters or it will
     * invalidate the event.
     */
    @objc
    public var interactionID: String?

    /**
     * The event's properties. Properties must be valid JSON.
     */
    @objc
    public var properties : [AnyHashable: Any]?

    @objc
    public var eventType : String {
        get {
            return "enhanced_custom_event"
        }
    }

    @objc
    public var priority: UAEventPriority {
        get {
            return .normal
        }
    }

    /**
     * Constructor.
     *
     * @param name The name of the event. The event's name must not exceed
     * 255 characters or it will invalidate the event.
     * @returns A Custom event instance
     */
    public init(name: String, value: NSNumber?) {
        self.eventName = name
        super.init()
        self.eventValue = value
    }

    /**
     * Constructor
     *
     * @param name The name of the event. The event's name must not exceed
     * 255 characters or it will invalidate the event.
     * @param stringValue The value of the event as a string. The value must be a valid
     * number between -2^31 and 2^31 - 1 or it will invalidate the event.
     * @returns A Custom event instance
     */
    public convenience init(name: String, stringValue: String?) {
        let decimalValue = stringValue != nil ? NSDecimalNumber(string: stringValue) : nil
        self.init(name: name, value: decimalValue)
    }

    /**
     * Constructor.
     *
     * @param name The name of the event. The event's name must not exceed
     * 255 characters or it will invalidate the event.
     */
    public convenience init(name:  String) {
        self.init(name: name, value: nil)
    }

    /**
     * Factory method for creating a custom event.
     *
     * @param name The name of the event. The event's name must not exceed
     * 255 characters or it will invalidate the event.
     * @returns A Custom event instance
     */
    @objc(eventWithName:)
    public class func event(name: String) -> UACustomEvent {
        return UACustomEvent(name: name)
    }

    /**
     * Factory method for creating a custom event with a value from a string.
     *
     * @param name The name of the event. The event's name must not exceed
     * 255 characters or it will invalidate the event.
     * @param string The value of the event as a string. The value must be a valid
     * number between -2^31 and 2^31 - 1 or it will invalidate the event.
     * @returns A Custom event instance
     */
    @objc(eventWithName:valueFromString:)
    public class func event(name: String, string: String?) -> UACustomEvent {
        return UACustomEvent(name: name, stringValue: string)
    }

    /**
     * Factory method for creating a custom event with a value.
     *
     * @param name The name of the event. The event's name must not exceed
     * 255 characters or it will invalidate the event.
     * @param value The value of the event. The value must be between -2^31 and
     * 2^31 - 1 or it will invalidate the event.
     * @returns A Custom event instance
     */
    @objc(eventWithName:value:)
    public class func event(name: String, value: NSNumber?) -> UACustomEvent {
        if (value == nil) {
            return UACustomEvent(name: name)
        } else if let decimal = value as? NSDecimalNumber  {
            return UACustomEvent(name: name, value: decimal)
        } else {
            let converted = NSDecimalNumber.init(decimal: value!.decimalValue)
            return UACustomEvent(name: name, value: converted)
        }
    }

    @objc
    public func isValid() -> Bool {
        var isValid = true
        isValid = self.isValid(string: self.eventName, name: "eventName", required: true) && isValid
        isValid = self.isValid(string: self.interactionType, name: "interactionType", required: false) && isValid
        isValid = self.isValid(string: self.interactionID, name: "interactionID", required: false) && isValid
        isValid = self.isValid(string: self.transactionID, name: "transactionID", required: false) && isValid
        isValid = self.isValid(string: self.templateType, name: "templateType", required: false) && isValid

        if let eventValue = self._eventValue {
            if eventValue == NSDecimalNumber.notANumber {
                AirshipLogger.error("Event value is not a number.")
                isValid = false
            } else if (eventValue.compare(NSNumber(value: Int32.max)).rawValue > 0) {
                AirshipLogger.error("Event value \(eventValue) is larger than 2^31-1.")
                isValid = false
            } else if (eventValue.compare(NSNumber(value: Int32.min)).rawValue < 0) {
                AirshipLogger.error("Event value \(eventValue) is smaller than -2^31.")
                isValid = false
            }
        }

        if let properties = self.properties {
            do {
                let propetyData = try JSONSerialization.data(withJSONObject: properties, options: [])
                if (propetyData.count > UACustomEvent.maxPropertiesSize) {
                    AirshipLogger.error("Event properties (%lu bytes) are larger than the maximum size of \(UACustomEvent.maxPropertiesSize) bytes.")
                    isValid = false
                }
            } catch {
                AirshipLogger.error("Event properties serialization error \(error)")
                isValid = false
            }
        }

        return isValid
    }

    /**
     * Sets the custom event's interaction type and identifier as coming from a Message Center message.
     * @param messageID The message ID.
     * @note For internal use only. :nodoc:
     */
    @objc
    public func setInteractionFromMessageCenterMessage(_ messageID: String?) {
        self.interactionID = messageID
        self.interactionType = UACustomEvent.interactionMCRAP
    }

    @objc
    public var data : [AnyHashable : Any] {
        get {
            let sendID = conversionSendID ?? UAirship.analytics()?.conversionSendID
            let sendMetadata = conversionPushMetadata ?? UAirship.analytics()?.conversionPushMetadata

            var dictionary: [AnyHashable : Any] = [:]
            dictionary[UACustomEvent.eventNameKey] = eventName
            dictionary[UACustomEvent.eventConversionSendIDKey] = sendID
            dictionary[UACustomEvent.eventConversionMetadataKey] = sendMetadata
            dictionary[UACustomEvent.eventInteractionIDKey] = interactionID
            dictionary[UACustomEvent.eventInteractionTypeKey] = interactionType
            dictionary[UACustomEvent.eventTransactionIDKey] = transactionID
            dictionary[UACustomEvent.eventTemplateTypeKey] = templateType
            dictionary[UACustomEvent.eventPropertiesKey] = properties

            if let eventValue = self._eventValue {
                let number = eventValue.multiplying(byPowerOf10: 6)
                dictionary[UACustomEvent.eventValueKey] = number.int64Value
            }

            return dictionary
        }
    }

    /**
     * @note For internal use only. :nodoc:
     */
    @objc
    public var payload : [AnyHashable : Any] {
        get {
            /*
             * We are unable to use the event.data for automation because we modify some
             * values to be stringified versions before we store the event to be sent to
             * warp9. Instead we are going to recreate the event data with the unmodified
             * values.
             */
            var eventData: [AnyHashable : Any] = [:]
            eventData[UACustomEvent.eventNameKey] = eventName
            eventData[UACustomEvent.eventInteractionIDKey] = interactionID
            eventData[UACustomEvent.eventInteractionTypeKey] = interactionType
            eventData[UACustomEvent.eventTransactionIDKey] = transactionID
            eventData[UACustomEvent.eventValueKey] = eventValue
            eventData[UACustomEvent.eventPropertiesKey] = properties
            return eventData
        }
    }

    /**
     * Adds the event to analytics.
     */
    @objc
    public func track() {
        UAirship.analytics()?.add(self)
    }

    private func isValid(string: String?, name: String, required: Bool = false) -> Bool {
        guard let string = string else {
            if (required) {
                AirshipLogger.error("Missing requied field \(name)")
                return false
            } else {
                return true
            }
        }

        guard (!required || string.count > 0) && string.count <= 255 else {
            AirshipLogger.error("Field \(name) must be between \(required ? 1 : 0) and 255 characters.")
            return false
        }

        return true
    }
}
